man/clone.8
.TH CLONE 8 "Mon Jul 20 2009" "" "Network UPS Tools (NUT)"
.SH NAME
clone \- UPS driver clone

.SH NOTE
This man page only documents the specific features of the
clone driver.  For information about the core driver, see
\fBnutupsdrv\fR(8).

.SH DESCRIPTION
This driver, which sits on top of another driver socket, allows to group
clients to a particular outlet of a device and deal with this output as
if it was a normal UPS.

.SH EXTRA ARGUMENTS
This driver supports the following settings:

.IP "load.off=\fIcommand\fR"
Recommended.  Set the command on the 'real' UPS driver that will be used to
switch off the outlet.  You need both \fIload.off\fR and \fIload.on\fR in
order to power cycle the outlet.  Otherwise, shutting down the clients
powered by an outlet is a one way street (see \fBIMPORTANT\fR).

.IP "load.on=\fIcommand\fR"
Recommended.  Set the command on the 'real' UPS driver that will be used to
switch on the outlet.  You need both \fIload.off\fR and \fIload.on\fR in
order to power cycle the outlet.  Otherwise, shutting down the clients
powered by an outlet is a one way street (see \fBIMPORTANT\fR).

.IP "load.status=\fIvalue\fR"
Recommended.  Set the variable on the 'real' UPS driver that will be used
to indicate the outlet status (ie on/off).  If not specified, the clone
driver will attempt to keep track of the outlet status, but this is less
reliable.

.IP "offdelay=\fInum\fR"
Set the timer (in seconds) before the outlet is turned off after the
shutdown condition (OB LB) for this outlet is met or a command to shutdown
was issued.  Defaults to 120 seconds.

.IP "ondelay=\fInum\fR"
Set the timer (in seconds) for the outlet to switch on in case the power
returns after the oulet has been switched off.  Defaults to 30 seconds.

.IP "mincharge=\fIvalue\fR"
Set the remaining battery level when the clone UPS switches to LB
(percent).

.IP "minruntime=\fIvalue\fR"
Set the remaining battery runtime when the clone UPS switches to LB
(seconds).

.SH IMPLEMENTATION
The port specification in the \fBups.conf\fR(5) reference the driver
socket that the 'real' UPS driver is using. For example:

.nf
[realups]
   driver = usbhid-ups
   port = auto
.fi

.nf
[clone-outlet-1]
   driver = clone
   port = usbhid-ups-realups
   load.on = outlet.1.load.on
   load.off = outlet.1.load.off
   load.status = outlet.1.status
   [...]
.fi

.SH IMPORTANT
Unlike a real UPS, you should \fBnot\fR configure a upsmon master for this
driver.  When a upsmon master sees the OB LB flags and tells the upsd server
it is OK to initiate the shutdown sequence, the server will latch the FSD
status and it will not be possible to restart the systems connected without
restarting the upsd server.

This will be a problem if the power returns after the clone UPS initiated
the shutdown sequence on it's outlet, but returns before the real UPS begins
shutting down. The solution is in the clone driver, that will insert the
FSD flag if needed without the help of a upsmon master.

.SH CAVEATS
The clone UPS will follow the status on the real UPS driver.  You can only
make the clone UPS shutdown earlier than the real UPS driver, not later.
If the real UPS driver initiates a shutdown, the clone UPS driver will
immediately follow.

Be aware that the commands to shutdown/restart an outlet on the real UPS
drivers are not affected, so if you tell the real UPS driver to shutdown
the outlet of the clone UPS driver, your clients will lose power without
warning.

.SH AUTHOR
Arjen de Korte <adkorte-guest@alioth.debian.org>

.SH SEE ALSO

\fBupscmd\fR(1),
\fBupsrw\fR(1),
\fBups.conf\fR(5),
\fBnutupsdrv\fR(8)

.SS Internet resources:
The NUT (Network UPS Tools) home page: http://www.networkupstools.org/


NEWS
If you're upgrading from an earlier version, see the UPGRADING file.

---------------------------------------------------------------------------
Release notes for NUT 2.4.1 - what's new since 2.4.0:

 - the microdowell driver has appeared to support various MicroDowell Enterprise
   units (see the "new devices" list below).

 - support for new devices: MicroDowell Enterprise B8, B10, N8, N11, N15, N20,
   N22, N30, N40, N50, N60 and HiBox ST.

 - NUT-Monitor now better handles the ups.status field, and has switched to
   version 1.1.

 - the situation of the build toolchain has been fixed, with regard to the
   "make clean" target and the wrongly removed generated USB files. This brokes
   further configure call. 

---------------------------------------------------------------------------
Release notes for NUT 2.4.0 - what's new since 2.2.2:

 - preliminary support for Power Distribution Units (PDUs): NUT now support
   PDUs, either natively (ie using NUT snmp-ups driver), or through a binding to
   the Powerman daemon. The list of supported PDUs is already quite long,
   including: Eaton ePDUs (Managed and Monitored), some Aphel models, some
   Raritan PDUs, and the whole list of Powerman supported devices:
   http://powerman.sourceforge.net/supported.html
 
 - support for new devices: the various PDUs cited above, Chloride Desk Power
   650, Cyber Power Systems Value 400E/600E/800E (USB models), Delta GES602N,
   Digitus DN-170020, the whole Eaton ranges (mostly composed of MGE Office
   Protection Systems and Powerware units) including BladeUPS, Forza Power
   Technologies SL-1001, HP PowerTrust 2997A, HP R/T 2200 G2, Infosec XP 1000
   and XP 500, Ippon Back Power Pro (serial and USB), Kebo 1200D/D Series,
   Liebert PowerSure Personal XT, MGE Office Protection Systems Protection
   Station, Neus 400va and 600va, Phasak 400VA and 600VA, Plexus 500VA, Powercom
   Black Knight PRO / King PRO and Imperial, PowerKinetics BlackOut Buster,
   Sweex 1000 USB, UNITEK Alpha 500, WinPower CPM-800.

 - NUT now embeds Python client support through the PyNUTClient module and the
   NUT-Monitor application. Both are from David Goncalves, and are still
   available from http://www.lestat.st.
   For more information, refer to scripts/python/README.

 - the dummy-ups driver now support a "repeater" mode. This allows it to act as
   a NUT client, and to forward data. This can be useful for supervision and
   load sharing purposes.

 - tcp-wrappers support has been added to the upsd server, to grant users access
   by source IP for commands that require to be logged into the server. This
   replaces the previous internal implementation (ACL in upsd.conf).

 - the nut.conf file has been introduced to standardize startup configuration
   across the various systems.

 - NUT now ships a bash completion function for 'upsc' command
   (scripts/misc/nut.bash_completion). Simply copy it to /etc/bash_completion.d
 
 - many internal changes to improve maintenability, while lowering the
   maintenance cost (thus allowing developers to focus on what matters: the
   code!). Examples of this are:
   - the USB information automatic extraction to generate the various USB helper
     files,
   - the upsdrv_info_t structure to track more driver information, and remove
     the need for the upsdrv_banner() function
   - common USB code refactoring, as it is done for the serial functions.

 - tons of bugfixes, cleanup and improvements to make NUT stronger than ever!

 For a complete list of changes, please refer to the ChangeLog file.

---------------------------------------------------------------------------
Release notes for NUT 2.2.2 - what's new since 2.2.1:

 - support for new devices: APC BACK-UPS XS LCD, Atlantis Land,
   Mustek Powermust Office 650, Oneac XAU models, Powerware PW5115 and
   PW9120 (USB), Nitram Elite 2005
 
 - Integrated Power Management (NUT HAL integration) has reached a
   major milestone: it is now the most advanced UPS integration into
   Power Management layer known in existing OSs. It has received many
   corrections and improvements, and allows to PowerOff the UPS at the
   end of a power cycle (which is the most important feature, not
   supported on other systems).
   The various files are now installed into the correct location.

 - the usbhid-ups driver has received attention. Most notably, the
   shutdown handling has been reworked, and support for MGE UPS SYSTEMS
   3 phases units has been added.

 - snmp-ups now supports MGE* Environment Sensor (ref 66 846).
   The ambient.temperature reporting has also been fixed for units
   other than APC.

 - the netxml-ups driver has appeared to support MGE* network HTTP/XML
   cards.
 
 - NUT now distributes by default the shared version of libupsclient
   (version 1.0.0), and use this for the provided clients (upsmon, upsc,
   upsrw, upscmd). This is part of an effort to reduce NUT's footprint,
   both on disk and in memory.
 
 - powerpanel has reach a new step toward the replacement of nitram and
   cpsups drivers. The final step is scheduled for NUT 2.4.
 
 - many changes, cleanup and fixes to the NUT core and various drivers.

 For a complete list of changes, please refer to the ChangeLog file.

---------------------------------------------------------------------------
Release notes for NUT 2.2.1 - what's new since 2.2.0:

 - support for new devices: all MGE Office Protection Systems units,
   Advice TopGuard 2000, Belkin F6H375-USB, Dynamix UPS1700D, Effekta RM2000MH,
   Jageson Technology Jasuny USPS, Powercom SMK-1500A and SXL-1500A,
   PowerWalker Line-Interactive VI 400/800 and 600, Powerware 9110,
   UNITEK Alpha 2600, UPSonic CXR1000, some vintage serial APC UPSs.

 - the usbhid-ups driver has been improved, and fixed in many areas, through
   a backport of the development (trunk) version.

 - the udev rules, for Linux hotplug support of the USB UPSs, has been
   updated to support kernel newer than 2.6.22.


drivers/clone.c
/*
* clone.c: UPS driver clone
*
* Copyright (C) 2009 - Arjen de Korte <adkorte-guest@alioth.debian.org>
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 2 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
*/

#include "main.h"
#include "parseconf.h"

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>

#define DRIVER_NAME	"Clone UPS driver"
#define DRIVER_VERSION	"0.02"

/* driver description structure */
upsdrv_info_t upsdrv_info = {
	DRIVER_NAME,
	DRIVER_VERSION,
	"Arjen de Korte <adkorte-guest@alioth.debian.org>",
	DRV_EXPERIMENTAL,
	{ NULL }
};

static struct {
	struct {
		int	start;
		int	shutdown;
	} timer;
	char	status[ST_MAX_VALUE_LEN];
} ups = { { -1, -1 }, "WAIT" };

static struct {
	struct {
		double	act;
		double	low;
	} charge;
	struct {
		double	act;
		double	low;
	} runtime;
} battery = { { 0, 0 }, { 0, 0 } };

static int	dumpdone = 0, online = 1, outlet = 1;
static int	offdelay = 120, ondelay = 30;

static PCONF_CTX_t	sock_ctx;
static time_t	last_poll = 0, last_heard = 0,
		last_ping = 0, last_connfail = 0;

static int instcmd(const char *cmdname, const char *extra);


static int parse_args(int numargs, char **arg)
{
	if (numargs < 1) {
		return 0;
	}

	if (!strcasecmp(arg[0], "PONG")) {
		upsdebugx(3, "Got PONG from UPS");
		return 1;
	}

	if (!strcasecmp(arg[0], "DUMPDONE")) {
		upsdebugx(3, "UPS: dump is done");
		dumpdone = 1;
		return 1;
	}

	if (!strcasecmp(arg[0], "DATASTALE")) {
		dstate_datastale();
		return 1;
	}

	if (!strcasecmp(arg[0], "DATAOK")) {
		dstate_dataok();
		return 1;
	}

	if (numargs < 2) {
		return 0;
	}

	/* DELINFO <var> */
	if (!strcasecmp(arg[0], "DELINFO")) {
		dstate_delinfo(arg[1]);
		return 1;
	}

	if (numargs < 3) {
		return 0;
	}

	/* SETINFO <varname> <value> */
	if (!strcasecmp(arg[0], "SETINFO")) {

		if (!strncasecmp(arg[1], "driver.", 7) ||
				!strcasecmp(arg[1], "battery.charge.low") ||
				!strcasecmp(arg[1], "battery.runtime.low") ||
				!strncasecmp(arg[1], "ups.delay.", 10) ||
				!strncasecmp(arg[1], "ups.timer.", 10)) {
			/* don't pass on upstream driver settings */
			return 1;
		}

		if (!strcasecmp(arg[1], "ups.status")) {
			snprintf(ups.status, sizeof(ups.status), "%s", arg[2]);

			online = strstr(ups.status, "OL") ? 1 : 0;

			if (ups.timer.shutdown > 0) {
				dstate_setinfo("ups.status", "FSD %s", ups.status);
				return 1;
			}
		}

		if (!strcasecmp(arg[1], "battery.charge")) {
			battery.charge.act = strtod(arg[2], NULL);

			dstate_setinfo("battery.charge.low", "%g", battery.charge.low);
			dstate_setflags("battery.charge.low", ST_FLAG_RW | ST_FLAG_STRING);
			dstate_setaux("battery.charge.low", 3);
		}

		if (!strcasecmp(arg[1], "battery.runtime")) {
			battery.runtime.act = strtod(arg[2], NULL);

			dstate_setinfo("battery.runtime.low", "%g", battery.runtime.low);
			dstate_setflags("battery.runtime.low", ST_FLAG_RW | ST_FLAG_STRING);
			dstate_setaux("battery.runtime.low", 4);
		}

		dstate_setinfo(arg[1], arg[2]);
		return 1;
	}

	return 0;
}


static int sstate_connect(void)
{
	int	ret, fd;
	const char	*dumpcmd = "DUMPALL\n";
	struct sockaddr_un	sa;

	memset(&sa, '\0', sizeof(sa));
	sa.sun_family = AF_UNIX;
	snprintf(sa.sun_path, sizeof(sa.sun_path), "%s/%s", dflt_statepath(), device_path);

	fd = socket(AF_UNIX, SOCK_STREAM, 0);

	if (fd < 0) {
		upslog_with_errno(LOG_ERR, "Can't create socket for UPS [%s]", device_path);
		return -1;
	}

	ret = connect(fd, (struct sockaddr *) &sa, sizeof(sa));

	if (ret < 0) {
		time_t	now;

		close(fd);

		/* rate-limit complaints - don't spam the syslog */
		time(&now);

		if (difftime(now, last_connfail) < 60) {
			return -1;
		}

		last_connfail = now;

		upslog_with_errno(LOG_ERR, "Can't connect to UPS [%s]", device_path);
		return -1;
	}

	ret = fcntl(fd, F_GETFL, 0);

	if (ret < 0) {
		upslog_with_errno(LOG_ERR, "fcntl get on UPS [%s] failed", device_path);
		close(fd);
		return -1;
	}

	ret = fcntl(fd, F_SETFL, ret | O_NDELAY);

	if (ret < 0) {
		upslog_with_errno(LOG_ERR, "fcntl set O_NDELAY on UPS [%s] failed", device_path);
		close(fd);
		return -1;
	}

	/* get a dump started so we have a fresh set of data */
	ret = write(fd, dumpcmd, strlen(dumpcmd));

	if (ret != (int)strlen(dumpcmd)) {
		upslog_with_errno(LOG_ERR, "Initial write to UPS [%s] failed", device_path);
		close(fd);
		return -1;
	}

	pconf_init(&sock_ctx, NULL);

	time(&last_heard);

	dumpdone = 0;

	/* set ups.status to "WAIT" while waiting for the driver response to dumpcmd */
	dstate_setinfo("ups.status", "WAIT");

	upslogx(LOG_INFO, "Connected to UPS [%s]", device_path);
	return fd;
}


static void sstate_disconnect(void)
{
	if (upsfd < 0) {
		return;
	}

	pconf_finish(&sock_ctx);

	close(upsfd);
	upsfd = -1;
}


static int sstate_sendline(const char *buf)
{
	int	ret;

	if (upsfd < 0) {
		return -1;	/* failed */
	}

	ret = write(upsfd, buf, strlen(buf));

	if (ret == (int)strlen(buf)) {
		return 0;
	}

	upslog_with_errno(LOG_NOTICE, "Send to UPS [%s] failed", device_path);
	return -1;	/* failed */
}


static int sstate_readline(void)
{
	int	i, ret;
	char	buf[SMALLBUF];

	if (upsfd < 0) {
		return -1;	/* failed */
	}

	ret = read(upsfd, buf, sizeof(buf));

	if (ret < 0) {
		switch(errno)
		{
		case EINTR:
		case EAGAIN:
			return 0;

		default:
			upslog_with_errno(LOG_WARNING, "Read from UPS [%s] failed", device_path);
			return -1;
		}
	}

	for (i = 0; i < ret; i++) {

		switch (pconf_char(&sock_ctx, buf[i]))
		{
		case 1:
			if (parse_args(sock_ctx.numargs, sock_ctx.arglist)) {
				time(&last_heard);
			}
			continue;

		case 0:
			continue;	/* haven't gotten a line yet */

		default:
			/* parse error */
			upslogx(LOG_NOTICE, "Parse error on sock: %s", sock_ctx.errmsg);
			return -1;
		}
	}

	return 0;
}


static int sstate_dead(int maxage)
{
	time_t	now;
	double	elapsed;

	/* an unconnected ups is always dead */
	if (upsfd < 0) {
		upsdebugx(3, "sstate_dead: connection to driver socket for UPS [%s] lost", device_path);
		return -1;	/* dead */
	}

	time(&now);

	/* ignore DATAOK/DATASTALE unless the dump is done */
	if (dumpdone && dstate_is_stale()) {
		upsdebugx(3, "sstate_dead: driver for UPS [%s] says data is stale", device_path);
		return -1;	/* dead */
	}

	elapsed = difftime(now, last_heard);

	/* somewhere beyond a third of the maximum time - prod it to make it talk */
	if ((elapsed > (maxage / 3)) && (difftime(now, last_ping) > (maxage / 3))) {
		upsdebugx(3, "Send PING to UPS");
		sstate_sendline("PING\n");
		last_ping = now;
	}

	if (elapsed > maxage) {
		upsdebugx(3, "sstate_dead: didn't hear from driver for UPS [%s] for %g seconds (max %d)",
			   device_path, elapsed, maxage);
		return -1;	/* dead */
	}

	return 0;
}


static int instcmd(const char *cmdname, const char *extra)
{
	const char	*val;

	val = dstate_getinfo(getval("load.status"));
	if (val) {
		if (!strcasecmp(val, "off") || !strcasecmp(val, "no")) {
			outlet = 0;
		}

		if (!strcasecmp(val, "on") || !strcasecmp(val, "yes")) {
			outlet = 1;
		}
	}

	if (!strcasecmp(cmdname, "shutdown.return")) {
		if (outlet && (ups.timer.shutdown < 0)) {
			ups.timer.shutdown = offdelay;
			dstate_setinfo("ups.status", "FSD %s", ups.status);
		}
		ups.timer.start = ondelay;
		return STAT_INSTCMD_HANDLED;
	}

	if (!strcasecmp(cmdname, "shutdown.stayoff")) {
		if (outlet && (ups.timer.shutdown < 0)) {
			ups.timer.shutdown = offdelay;
			dstate_setinfo("ups.status", "FSD %s", ups.status);
		}
		ups.timer.start = -1;
		return STAT_INSTCMD_HANDLED;
	}

	upslogx(LOG_NOTICE, "instcmd: unknown command [%s]", cmdname);
	return STAT_INSTCMD_UNKNOWN;
}


static int setvar(const char *varname, const char *val)
{
	if (!strcasecmp(varname, "battery.charge.low")) {
		battery.charge.low = strtod(val, NULL);
		dstate_setinfo("battery.charge.low", "%g", battery.charge.low);
		return STAT_SET_HANDLED;
	}

	if (!strcasecmp(varname, "battery.runtime.low")) {
		battery.runtime.low = strtod(val, NULL);
		dstate_setinfo("battery.runtime.low", "%g", battery.runtime.low);
		return STAT_SET_HANDLED;
	}

	upslogx(LOG_NOTICE, "setvar: unknown variable [%s]", varname);
	return STAT_SET_UNKNOWN;
}


void upsdrv_initinfo(void)
{
	const char	*val;

	time(&last_poll);

	dstate_addcmd("shutdown.return");
	dstate_addcmd("shutdown.stayoff");

	val = getval("offdelay");
	if (val) {
		offdelay = strtol(val, NULL, 10);
	}

	val = getval("ondelay");
	if (val) {
		ondelay = strtol(val, NULL, 10);
	}

	val = getval("mincharge");
	if (val) {
		battery.charge.low = strtod(val, NULL);
	}

	val = getval("minruntime");
	if (val) {
		battery.runtime.low = strtod(val, NULL);
	}

	dstate_setinfo("ups.delay.shutdown", "%d", offdelay);
	dstate_setinfo("ups.delay.start", "%d", ondelay);

	dstate_setinfo("ups.timer.shutdown", "%d", ups.timer.shutdown);
	dstate_setinfo("ups.timer.start", "%d", ups.timer.start);

	upsh.instcmd = instcmd;
	upsh.setvar = setvar;
}


void upsdrv_updateinfo(void)
{
	time_t	now = time(NULL);

	if (sstate_dead(15)) {
		sstate_disconnect();
		extrafd = upsfd = sstate_connect();
		return;
	}

	if (sstate_readline()) {
		sstate_disconnect();
		return;
	}

	if (ups.timer.shutdown >= 0) {

		ups.timer.shutdown -= difftime(now, last_poll);

		if (ups.timer.shutdown < 0) {
			const char	*val;

			ups.timer.shutdown = -1;
			outlet = 0;

			val = getval("load.off");
			if (val) {
				char	buf[SMALLBUF];
				snprintf(buf, sizeof(buf), "INSTCMD %s\n", val);
				sstate_sendline(buf);
			}
		}

	} else if (ups.timer.start >= 0) {

		if (online) {
			ups.timer.start -= difftime(now, last_poll);
		} else {
			ups.timer.start = ondelay;
		}

		if (ups.timer.start < 0) {
			const char	*val;

			ups.timer.start = -1;
			outlet = 1;

			val = getval("load.on");
			if (val) {
				char	buf[SMALLBUF];
				snprintf(buf, sizeof(buf), "INSTCMD %s\n", val);
				sstate_sendline(buf);
			}

			dstate_setinfo("ups.status", "%s", ups.status);
		}

	} else if (!online && outlet) {

		if (battery.charge.act < battery.charge.low) {
			upslogx(LOG_INFO, "Battery charge low");
			instcmd("shutdown.return", NULL);
		} else if (battery.runtime.act < battery.runtime.low) {
			upslogx(LOG_INFO, "Battery runtime low");
			instcmd("shutdown.return", NULL);
		}
	}

	dstate_setinfo("ups.timer.shutdown", "%d", ups.timer.shutdown);
	dstate_setinfo("ups.timer.start", "%d", ups.timer.start);

	last_poll = now;
}


void upsdrv_shutdown(void)
{
	fatalx(EXIT_FAILURE, "shutdown not supported");
}


void upsdrv_help(void)
{
}


void upsdrv_makevartable(void)
{
	addvar(VAR_VALUE, "offdelay", "Delay before outlet shutdown (seconds)");
	addvar(VAR_VALUE, "ondelay", "Delay before outlet startup (seconds)");

	addvar(VAR_VALUE, "mincharge", "Remaining battery level when UPS switches to LB (percent)");
	addvar(VAR_VALUE, "minruntime", "Remaining battery runtime when UPS switches to LB (seconds)");

	addvar(VAR_VALUE, "load.off", "Command to switch off outlet");
	addvar(VAR_VALUE, "load.on", "Command to switch on outlet");
	addvar(VAR_VALUE, "load.status", "Variable that indicates outlet is on/off");
}


void upsdrv_initups(void)
{
	extrafd = upsfd = sstate_connect();
}


void upsdrv_cleanup(void)
{
	sstate_disconnect();
}
